rest-api/dao/biobank_order_dao.py

```diff
@@ -339,4 +391,140 @@ def to_client_json(self, model):
+      order.restoredSiteId = get_site(resource['restoredInfo'])
+      order.restoredTime = clock.CLOCK.now()
+      order.orderStatus = BiobankOrderStatus.UNSET
+    else:
+      raise BadRequest('status must be restored or cancelled for patch request.')
+
+    self._update_history(session, order)
+    self._update_identifier_history(session, order)
+    self._update_sample_history(session, order)
+    super(BiobankOrderDao, self)._do_update(session, order, resource)
+    return order
+
+  def _validate_patch_update(self, session, model, resource, expected_version):
+    required_cancelled_fields = ['amendedReason', 'cancelledInfo', 'status']
+    required_restored_fields = ['amendedReason', 'restoredInfo', 'status']
+    if 'status' not in resource:
+      raise BadRequest('status of cancelled/restored is required')
+
+    if resource['status'] == 'cancelled':
+      for field in required_cancelled_fields:
+        if field not in resource:
+          raise BadRequest('%s is required for a cancelled biobank order' % field)
+      if 'site' not in resource['cancelledInfo'] or 'author' not in resource['cancelledInfo']:
+        raise BadRequest('author and site are required for cancelledInfo')
+
+    elif resource['status'] == 'restored':
+      for field in required_restored_fields:
+        if field not in resource:
+          raise BadRequest('%s is required for a cancelled biobank order' % field)
+      if 'site' not in resource['restoredInfo'] or 'author' not in resource['restoredInfo']:
+        raise BadRequest('author and site are required for restoredInfo')
+
+    super(BiobankOrderDao, self)._validate_patch_update(session, model, resource, expected_version)
+
+  @staticmethod
+  def _update_history(session, order):
+    # Increment the version and add a new history entry.
+    session.flush()
+    history = BiobankOrderHistory()
+    history.fromdict(order.asdict(follow=['logPosition']), allow_pk=True)
+    history.logPositionId = order.logPosition.logPositionId
+    session.add(history)
+
+  @staticmethod
+  def _update_identifier_history(session, order):
+    session.flush()
+    for identifier in order.identifiers:
+      history = BiobankOrderIdentifierHistory()
+      history.fromdict(identifier.asdict(), allow_pk=True)
+      history.version = order.version
+      history.biobankOrderId = order.biobankOrderId
+      session.add(history)
+
+  @staticmethod
+  def _update_sample_history(session, order):
+    session.flush()
+    for sample in order.samples:
+      history = BiobankOrderedSampleHistory()
+      history.fromdict(sample.asdict(), allow_pk=True)
+      history.version = order.version
+      history.biobankOrderId = order.biobankOrderId
+      session.add(history)
+
+  @staticmethod
+  def _clear_cancelled_and_restored_fields(order):
+    #pylint: disable=unused-argument
+    """ Just in case these fields have values, we don't want them in the most recent record,
```

Comment:
I don't know if this clearing is strictly necessary; having the latest row maintain the users, timestamp, and site when an order was restored, cancelled, etc. seems OK, too. (We'll have the status field to tell us what state the order is in.)